client/block-builder-ver/src/lib.rs

```diff
@@ -223,17 +221,84 @@ where
 			api,
 			block_id,
 			backend,
-			previous_block_extrinsics: None,
 			estimated_header_size,
 		})
 	}
 
 	/// temporaily apply extrinsics and record them on the list
-	pub fn build_with_seed<F: FnOnce(&'_ BlockId<Block>, &'_ A::Api) -> Vec<Block::Extrinsic>>(
+	pub fn build_with_seed<
+		F: FnOnce(
+			&'_ BlockId<Block>,
+			&'_ A::Api,
+		) -> Vec<(Option<sp_runtime::AccountId32>, Block::Extrinsic)>,
+	>(
 		mut self,
 		seed: ShufflingSeed,
 		call: F,
 	) -> Result<BuiltBlock<Block, backend::StateBackendFor<B, Block>>, Error> {
+		let block_id = self.block_id;
+
+		let mut valid_txs = if self.api.can_enqueue_txs(&block_id).unwrap() {
+			self.api.execute_in_transaction(|api| {
+				let next_header = api
+					.finalize_block_with_context(&block_id, ExecutionContext::BlockConstruction)
+					.unwrap();
+
+				api.start_prevalidation(&block_id).unwrap();
+
+				// create dummy header just to condider N+1 block extrinsics like new session
+				let header = <<Block as BlockT>::Header as HeaderT>::new(
+					*next_header.number() + One::one(),
+					Default::default(),
+					Default::default(),
+					next_header.hash(),
+					Default::default(),
+				);
+
+				if api.is_storage_migration_scheduled(&self.block_id).unwrap() {
+					log::debug!(target:"block_builder", "storage migration scheduled - ignoring any txs");
+					TransactionOutcome::Rollback(vec![])
+				} else {
+					api.initialize_block_with_context(
+						&self.block_id,
+						ExecutionContext::BlockConstruction,
+						&header,
+					)
+					.unwrap();
+					let txs = call(&self.block_id, &api);
+					TransactionOutcome::Rollback(txs)
+				}
+			})
+		} else {
+			log::info!(target:"block_builder", "storage queue is full, no room for new txs");
+			vec![]
+		};
+
+		// that should be improved at some point
+		if valid_txs.len() > 100 {
+			valid_txs.truncate(valid_txs.len() * 90 / 100);
+		}
+
+		let valid_txs_count = valid_txs.len();
+		let store_txs_inherent = self
+			.api
+			.create_enqueue_txs_inherent(
+				&self.block_id,
+				valid_txs.into_iter().map(|(_, tx)| tx).collect(),
+			)
+			.unwrap();
+
+		apply_transaction_wrapper::<Block, A>(
+			&self.api,
+			&self.block_id,
+			store_txs_inherent.clone(),
+			ExecutionContext::BlockConstruction,
+		)
+		.unwrap()
+		.unwrap()
+		.unwrap();
```

Comment:
put all txs collected from transaction pool into inherent that stores them in storage queue